### ISSUE TITLE
False negative with useless-expression on strings

### ISSUE THREAD

**Issue by NeilGirdhar** [astral-sh/ruff#11292]
For some reason useless-expression is not triggered by loose strings.  Ideally, it should catch this and suggest that they be converted to comments if that's the intention.
```python
class X:
    """Some docstring."""


def f() -> None:
    """Some docstring."""
    [123]  # noqa: B018
    "Some loose string."  # Should be B018.


x = 1
"Some editors consider this a variable docstring."
print(X.__doc__, f.__doc__)  # noqa: T201
```

**Comment by charliermarsh** [comment]
The intention here is to avoid flagging docstrings, but I think we need to make it more targeted.

**Comment by NeilGirdhar** [comment]
Yeah, that's what I figured.  I tried to illustrate some reasonable uses of strings.  I'm not sure how you feel about the "variable docstring", which some editors recognize.  As far as I know, they're rarely used and not really a Python concept.  They look pretty weird in code:
```python
length: int  # The length of the object.
width: int  # The width of the object.
```
versus
```
length: int
"The length of the object."
width: int
"The width of the object."
```

**Comment by charliermarsh** [comment]
Yeah, I think we _do_ need to ignore attribute docstrings. They're not standardized, but they're popular enough that I think it'd be a net-negative to flag them.

**Comment by dhruvmanila** [comment]
I think this can be fixed because our docstring detection is quite accurate now.

**Comment by dhruvmanila** [comment]
It seems that updating `B018` seems to raise a lot of diagnostics in the ecosystem checks. What about implementing the [`PLW0105`](https://pylint.pycqa.org/en/latest/user_guide/messages/warning/pointless-string-statement.html) rule from Pylint instead of updating `B018`? This will also match `flake8-bugbear` behavior.

**Comment by NeilGirdhar** [comment]
Please don't use the same documentation as PLW0105: strings are not reasonable documentation in Python, and I don't think the linter should suggest that.

**Comment by estyrke** [comment]
If B018 is to be modified, consider the special case of useless string expression after a string assignment. This is rather nasty and hard to detect for a human (the correct code should have parentheses to make this a multi line implicit string concatenation):

```python
    def __str__(self) -> str:
        string = f"AnalysisConfig(ID: {self.uuid},"
        f"BackendURL: {self.backend_url},"
        f"AnnotationTypes: {self.annotation_types})"
        return string
```

**Comment by NeilGirdhar** [comment]
Maybe Ruff should flag "attribute docstrings" using a separate rule to convert them to annotated annotations?
```
x: int
"blah"
```
becomes
```
x: Annotated[int, "blah"]
```
This is more consistent with modern Python.

**Comment by MichaReiser** [comment]
This has come up in ty (https://github.com/astral-sh/ty/issues/4204) and I was also surprised that `useless-expression` does not catch this.

**Comment by MichaReiser** [comment]
Let me create a PR for this to understand the ecosystem impact

**Comment by ntBre** [comment]
Dhruv has an old PR with some discussion. I came across it recently and wondered if we should finish it off too. Ah it was for https://github.com/astral-sh/ruff/pull/27253, the ecosystem check there might help too?

**Comment by MichaReiser** [comment]
Thanks. I saw @dhruvmanila PR, but I considered creating a new PR might be easier at this point than rebasing a 2 year old PR 😆 

I wasn't aware of #27253 but I agree with your sentiment that changing this rule makes more sense

**Comment by yonatangross** [comment]
The 2024 ecosystem noise is the thing worth splitting before this run.

dhruv's 2024-06-03 note ("a lot of diagnostics") is what pushed this toward PLW0105 as a separate rule. But the strings B018 would newly flag are not one population, and the two halves have opposite risk:

```python
x: int
"Attribute docstring."        # string after an assignment or annotation, at
                              # class/module level. @charliermarsh already
                              # ruled these out as a net-negative to flag.

def f() -> None:
    """Real docstring."""
    "Silently dropped."       # string expression in a function body that is
                              # not the body's first statement. Cannot reach
                              # __doc__ by any path.
```

The second case is what I hit in ty#4204 (a string between `@overload` definitions, where mypy's overload collection then truncates the series, so it is not purely cosmetic there). It is structurally disjoint from attribute docstrings: different parent node, different position.

Concrete ask for the ecosystem run: bucket the new diagnostics by parent node type rather than reading one total. If the volume is dominated by the assignment-adjacent bucket, a B018 change scoped to "string expression in a function body, not the first statement" may land with a small delta and no new rule needed.

I have not measured this, so it is a hypothesis your run would settle either way. Happy to do the bucketing myself if that is useful.

**Comment by MichaReiser** [comment]
This is a little more work because our attribute docstring detection doesn't support attributes in `__init__` or documenting type alises
